[MAINT] removed --py3k argument for pylint

orbeckst opened the pull request:
Fixes #3422

Changes made in this Pull Request:
- removed `pylint --py3k` option from GH CI workflow (pylint is not used in Azure or Travis) 
- py3k checker was removed in pylint 2.11
  https://pylint.pycqa.org/en/latest/whatsnew/2.11.html#removed-checkers


PR Checklist
------------
 - n/a Tests?
 - n/a Docs?
 - n/a CHANGELOG updated?
 - [x] Issue raised/referenced?

orbeckst committed:
[MAINT] removed --py3k argument for pylint

- fix #3422
- py3k checker was removed in pylint 2.11
  https://pylint.pycqa.org/en/latest/whatsnew/2.11.html#removed-checkers

IAlibay commented:
That's a lot fewer lint failures than I expected.

edit: at least RDKit we know for sure is fine (so we'll need to get pylint to ignore the global statement)

orbeckst commented:
If you have an idea how to fix/ignore the three issues then we can do it in this PR and start with a clean pylint.

IAlibay committed:
fix all pylint warnings

- ignore warning for justified global use in RDKit.py
- fix pylint W0233 warning
- removed deprecated "import imp" (not needed for Py > 3.3)
- disable pylint E1133 (not-an-iterable) on one line

pep8speaks commented:
Hello @orbeckst! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:











There are currently no PEP 8 issues detected in this Pull Request. Cheers! :beers: 

##### Comment last updated at 2021-10-05 11:19:25 UTC

IAlibay commented on package/MDAnalysis/coordinates/null.py:
```diff
@@ -60,4 +60,3 @@ def _write_next_frame(self, obj):
         except AttributeError:
             errmsg = "Input obj is neither an AtomGroup or Universe"
             raise TypeError(errmsg) from None
-        pass
```
Doesn't seem like this was required anyways?

orbeckst replied:
yep

IAlibay commented on package/MDAnalysis/converters/RDKit.py:
```diff
@@ -452,7 +452,7 @@ def set_converter_cache_size(maxsize):
         to the cache size, i.e. everything is cached.
     """
-    global atomgroup_to_mol
+    global atomgroup_to_mol # pylint: disable=global-statement
```
I believe we said this was a justified global call.

orbeckst replied:
I think so... caches are iffy.

IAlibay reviewed (commented):
Note sure about the third one.

Don't really understand why pylint is complaining about: https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/core/topologyattrs.py#L2346

IAlibay commented:
If the `type.__init__` call is fine, we can just add `# pylint: disable=W0233` on that line and it'll just ignore the warning.

orbeckst commented:
[W0233](https://pylint.pycqa.org/en/latest/technical_reference/features.html?highlight=W0233#classes-checker-messages): Used when an `__init__` method is called on a class which is not in the direct ancestors for the analysed class.

This metaclass `_ConnectionTopologyAttrMeta(_TopologyAttrMeta)` is ultimately derived from `type` https://github.com/MDAnalysis/mdanalysis/blob/39850215eae5a3e2c9d28a24dbc670646e305f79/package/MDAnalysis/core/topologyattrs.py#L287 so I am not quite sure why it triggers.

Who wrote that metaclass magic and might be able to comment? Is there a more pythonic way to do this?

orbeckst commented:
From blame, the most likely authors are @richardjgowers and @lilyminium — would you mind having a quick look here and comment on why pylint is complaining about the line 
```python
class _ConnectionTopologyAttrMeta(_TopologyAttrMeta):
    def __init__(cls, name, bases, classdict):
        type.__init__(type, name, bases, classdict)     # <---- pylint W0233
        attrname = classdict.get('attrname')

```
see https://github.com/MDAnalysis/mdanalysis/blob/39850215eae5a3e2c9d28a24dbc670646e305f79/package/MDAnalysis/core/topologyattrs.py#L2346

Thanks.

IAlibay commented:
> [W0233](https://pylint.pycqa.org/en/latest/technical_reference/features.html?highlight=W0233#classes-checker-messages): Used when an `__init__` method is called on a class which is not in the direct ancestors for the analysed class.
> 
> This metaclass `_ConnectionTopologyAttrMeta(_TopologyAttrMeta)` is ultimately derived from `type` https://github.com/MDAnalysis/mdanalysis/blob/39850215eae5a3e2c9d28a24dbc670646e305f79/package/MDAnalysis/core/topologyattrs.py#L287 so I am not quite sure why it triggers.
> 
> Who wrote that metaclass magic and might be able to comment? Is there a more pythonic way to do this?
> 
> 

@lilyminium I believe? If it works as intended we can just ignore it ?

lilyminium commented:
I think you could just remove `type.__init__`.

orbeckst commented:
ok, will try

orbeckst commented:
Looking good locally
```
$ pytest -n 4 --disable-pytest-warnings --durations=50 testsuite/MDAnalysisTests
======================================================= test session starts =======================================================
platform darwin -- Python 3.7.10, pytest-6.2.5, py-1.10.0, pluggy-1.0.0
...
====================== 17187 passed, 107 skipped, 2 xfailed, 2 xpassed, 45878 warnings in 432.02s (0:07:12) =======================
```

IAlibay commented on package/MDAnalysis/core/topologyattrs.py:
```diff
@@ -2343,7 +2343,6 @@ class _ConnectionTopologyAttrMeta(_TopologyAttrMeta):
     to return only the connections within the atoms in the group.
     """
     def __init__(cls, name, bases, classdict):
-        type.__init__(type, name, bases, classdict)
```
I guess the call just gets inherited from the parent class anyways (given we are calling super below)?

@richardjgowers - looks like you're the original author of the parent class, does it all make sense to you?

richardjgowers replied:
This seems to work

IAlibay reviewed (approved):
lgtm

IAlibay commented:
@richardjgowers - please review here, we need to get this merged so that CI stops being in failed status.

orbeckst commented:
ping @richardjgowers please!

richardjgowers committed:
I think this order is clearer

connectionmeta is just adding more stuff onto the existing meta, and
this way round expresses that

orbeckst commented:
Thanks @IAlibay and @richardjgowers !
